src/Tools/Pagination/Paginator.php

```diff
@@ -38,14 +40,43 @@ class Paginator implements Countable, IteratorAggregate
     private readonly Query $query;
     private bool|null $useOutputWalkers = null;
     private int|null $count             = null;
+    private bool $fetchJoinCollection;
+    /**
+     * @var bool The auto-detection of queries style was added a lot later to this class, and this
+     *  class historically was by default using the more complex queries style, which means that
+     *  the simple queries style is potentially very under-tested in production systems. The purpose
+     *  of this variable is to not introduce breaking changes until an impression is developed that
+     *  the simple queries style has been battle-tested enough.
+     */
+    private bool $queryStyleAutoDetectionEnabled = false;
+    /** @var bool|null Null means "undetermined". */
+    private bool|null $queryHasHavingClause = null;
 
-    /** @param bool $fetchJoinCollection Whether the query joins a collection (true by default). */
+    /**
+     * @param bool|null $fetchJoinCollection Whether the query joins a collection (true by default). Set
+     *  to null to enable auto-detection of this parameter, however note that the auto-detection requires
+     *  a QueryBuilder to be provided to Paginator, otherwise this parameter goes back to its default value.
+     *  Also, for now, when the auto-detection of this parameter is enabled, then auto-detection
+     *  of $useOutputWalkers and of the CountWalker::HINT_DISTINCT is also enabled.
+     */
     public function __construct(
         Query|QueryBuilder $query,
-        private readonly bool $fetchJoinCollection = true,
+        bool|null $fetchJoinCollection = true,
```

Comment:
I would revert this and instead introduce a static method.

```php
public static function newWithAutoDetection(QueryBuilder $queryBuilder): self
{
        // do the autodetection here
        // use the unaltered constructor here
        
}
```

That way: 
- you remove the possibility of passing a `Query` instance and `null` to the constructor.
- you don't complexify the constructor that much
- you don't rely on `null` for communicating failure

Reply:
Good idea. I've introduced the static function as you advised and the constructor returned to its original form in its entirety. Thanks.